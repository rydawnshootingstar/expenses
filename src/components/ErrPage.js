import React from 'react';

const ErrPage = () => (
    <div>
        404 not found dawg
        <br />
    </div>
);

export default ErrPage;