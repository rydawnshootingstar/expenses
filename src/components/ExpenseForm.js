import React from 'react';
import moment from 'moment';
import 'react-dates/initialize';
import {SingleDatePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

export default class ExpenseForm extends React.Component {

    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
    
        }
    }


    onDescriptionChange = (e)=>{
        const description = e.target.value;
        this.setState(()=> ({description}))
    };

    onNoteChange = (e)=> {
        const note = e.target.value;
        this.setState(()=> ({note}));
    };

    //doing it this way instead of checking the value on a user submit means that a user literally cannot
    //enter a value that does not match our specifications.
    onAmountChange = (e)=> {
        const amount = e.target.value;
        if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){            //added !amount to allow users to clear the field
            this.setState(()=> ({amount}));
        }
    };

    onDateChange = (createdAt)=> {
        if(createdAt){
            this.setState(()=>({createdAt}));
        }
    };

    onFocusChange = ({focused})=> {
        this.setState(()=> ({calendarFocused: focused}));
    };

    onSubmit = (e)=> {
        e.preventDefault();
        if(!this.state.description || !this.state.amount){
            //set error state 
            this.setState(()=>({error: 'description and amount fields are required!'}));
        }else{
            this.setState(()=> ({error:''}));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10)*100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            });
        }
    };

    render(){
        return (
            <div>
            {this.state.error!=='' && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input type="text" 
                    placeholder="Description" 
                    autoFocus 
                    value={this.state.description} 
                    onChange={this.onDescriptionChange}>
                    </input>

                    <input type="text" 
                    placeholder="Amount"
                    value={this.state.amount}
                    onChange={this.onAmountChange}
                    >
                    </input>

                    <SingleDatePicker
                    date={this.state.createdAt} // momentPropTypes.momentObj or null - cmon dawg I got moments
                    onDateChange={this.onDateChange} // PropTypes.func.isRequired
                    focused={this.state.calendarFocused} // PropTypes.bool
                    onFocusChange={this.onFocusChange} // PropTypes.func.isRequired
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    id="your_unique_id" // PropTypes.string.isRequired,
                    />

                    <textarea 
                    placeholder="Add a note for your expense (optional)"
                    value={this.state.note}
                    onChange={this.onNoteChange}
                    >
                    </textarea>

                    <button>Add Expense</button>
                </form>
            </div>
        );  
    }
}



