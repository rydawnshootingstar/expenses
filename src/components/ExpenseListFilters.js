import React from 'react';
import {connect} from 'react-redux';
import 'react-dates/initialize';
import {setTextFilter, sortByDate, sortByAmount, setStartDate, setEndDate} from '../actions/filters';
import {DateRangePicker} from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';



class ExpenseListFilters extends React.Component{

    constructor(props){
        super(props);
        this.state={
            calendarFocused: null
        }
    }

    onDatesChange=({startDate, endDate})=> {
        this.props.dispatch(setStartDate(startDate));
        this.props.dispatch(setEndDate(endDate));
    };

    onFocusChange=(calendarFocused)=>{
        this.setState(()=> ({calendarFocused}));
    };
    

    render (){ 
    return(
        <div className="content-container">
        <div className="input-group">
            <div className="input-group__item">
                <input 
                    className="text-input"
                    type="text" 
                    placeholder="Search expenses"
                    value={this.props.filters.text} 
                    onChange={(e)=>{
                        this.props.dispatch(setTextFilter(e.target.value));
                    }}>
                </input>
            </div>
            <div className="input-group__item">
                <select 
                    className="select"
                    onChange={(e)=> {
                        if(e.target.value === 'date'){
                            this.props.dispatch(sortByDate());
                        }else if(e.target.value ==='amount'){
                            this.props.dispatch(sortByAmount());
                        }
                    }}>
                        <option value="date">Date</option>
                        <option value="amount">Amount</option>
                </select>
            </div>
            <div className="input-group__item">
                <DateRangePicker
                    startDate={this.props.filters.startDate}
                    startDateId="date_from"
                    endDate={this.props.filters.endDate}
                    endDateId="date_to"
                    onDatesChange={this.onDatesChange}
                    focusedInput={this.state.calendarFocused}
                    onFocusChange={this.onFocusChange}
                    numberOfMonths={1}
                    isOutsideRange={()=>false}
                    showClearDates={true}
                />
            </div>
        </div>
    </div>
    )};
}

const mapStateToProps = (state)=> {
    return {
        filters: state.filters
    };
};

//when we connect to our redux store, we don't just get store.filters, but store.dispatch. In order to use any specific actions
//we must import them
export default connect(mapStateToProps)(ExpenseListFilters);


