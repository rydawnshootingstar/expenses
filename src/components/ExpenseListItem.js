import React from 'react'
import {connect} from 'react-redux';
import {removeExpense} from '../actions/expenses';
import {Link} from 'react-router-dom';


const ExpenseListItem = ({id, description, amount, createdAt})=> (
    <div>
    <Link to={`/edit/${id}`}><h3>{description}</h3></Link>
        <p>Cost: {amount}</p>
        <p>Created At: {createdAt}</p>
    </div>
);


export default (ExpenseListItem);