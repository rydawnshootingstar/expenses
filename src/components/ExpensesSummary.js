import React from 'react'
import {connect} from 'react-redux';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expensesTotal';
import numeral from 'numeral';
import {Link} from 'react-router-dom';

export const ExpensesSummary = ({count, total})=> {
const word = count === 1 ? 'expense' : 'expenses';
const finalTotal = numeral(total/ 100).format('$0,0.00');
    return(
        <div className="page-header">
        <div className="content-container">
            <h1 className="page-header__title">Currently viewing <span>{count}</span> {word} totalling <span>{finalTotal}</span></h1>
                <div className="page-header__actions">
                    <Link to="/create" className="button">Add Expense</Link>
                </div>
        </div>
        </div>
    );
};

const mapStateToProps = (state)=> {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    const expensesTotal = selectExpensesTotal(visibleExpenses);

    return {
        count: visibleExpenses.length,
        total: expensesTotal
    };
};
export default connect(mapStateToProps)(ExpensesSummary);