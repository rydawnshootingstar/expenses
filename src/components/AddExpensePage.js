import React from 'react';
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expenses';

//passes onsubmit prop to child so child can call it. This allows reusability when we go on to EditExpensePage
const AddExpensePage = (props) => (
    <div>
        <h1>Add Expense</h1>
        <ExpenseForm onSubmit={(expense)=> {
            props.dispatch(addExpense(expense));
            props.history.push('/');               //this is how we can redirect
        }}/>
    </div>
);

export default connect()(AddExpensePage);