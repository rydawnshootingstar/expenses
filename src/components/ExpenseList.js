import React from 'react'
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

export const ExpenseList = (props) => (
    <div>
        <h1>Expense List</h1>
        <p>
        {
            props.expenses.map((expense)=> {
                return <ExpenseListItem key={expense.id}{...expense} />
            })
        }
        </p>
    </div>
);

const mapStateToProps = (state)=> {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    };
};

//information about what we want to connect(state) as defined above, then the component
export default connect(mapStateToProps)(ExpenseList);