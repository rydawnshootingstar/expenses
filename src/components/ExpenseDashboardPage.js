import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpensesSummary from './ExpensesSummary';

const ExpenseDashboardPage = () => (
    <div>
        <ExpenseList />
        <ExpenseListFilters />
        <ExpensesSummary />
    </div>
);

export default ExpenseDashboardPage;