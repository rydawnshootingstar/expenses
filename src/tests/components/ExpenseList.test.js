import {ExpenseList} from '../../components/ExpenseList';
import {shallow} from 'enzyme';
import React from 'react';
import toJSON from 'enzyme-to-json';
import {testExpenses}from '../fixtures';

test('should render an expense list', ()=> {
    const wrapper = shallow(<ExpenseList expenses={testExpenses}/>);

    expect(toJSON(wrapper)).toMatchSnapshot();
});