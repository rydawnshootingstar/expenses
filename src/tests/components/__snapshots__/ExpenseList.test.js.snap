// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render an expense list 1`] = `
<div>
  <h1>
    Expense List
  </h1>
  <p>
    <ExpenseListItem
      amount={111}
      createdAt={0}
      description="exp 1"
      id="1"
      key="1"
      notes=""
    />
    <ExpenseListItem
      amount={222}
      createdAt={345600000}
      description="exp 2"
      id="2"
      key="2"
      notes=""
    />
    <ExpenseListItem
      amount={333}
      createdAt={3333}
      description="aaaa"
      id="3"
      key="3"
      notes=""
    />
    <ExpenseListItem
      amount={0}
      createdAt={-345600000}
      description="aaab"
      id="4"
      key="4"
      notes=""
    />
  </p>
</div>
`;
